### PR TITLE
Fix run make tests when it can't find dynamically linked librustc

### PR DESCRIPTION
**Pull request by aditijannu**
The recent updates to run make support caused the following tests to break on sgx i.e. target  `x86_64-fortanix-unknown-sgx` : 
```
11:22:48 failures:
11:22:48     [run-make] tests/run-make/CURRENT_RUSTC_VERSION
11:22:48     [run-make] tests/run-make/a-b-a-linker-guard
11:22:48     [run-make] tests/run-make/rustdoc-test-args
11:22:48     [run-make] tests/run-make/wasm-abi
11:22:48     [run-make] tests/run-make/wasm-custom-section
11:22:48     [run-make] tests/run-make/wasm-custom-sections-opt
11:22:48     [run-make] tests/run-make/wasm-export-all-symbols
11:22:48     [run-make] tests/run-make/wasm-import-module
11:22:48     [run-make] tests/run-make/wasm-panic-small
11:22:48     [run-make] tests/run-make/wasm-spurious-import
11:22:48     [run-make] tests/run-make/wasm-stringify-ints-small
11:22:48     [run-make] tests/run-make/wasm-symbols-different-module
11:22:48     [run-make] tests/run-make/wasm-symbols-not-exported
11:22:48     [run-make] tests/run-make/wasm-symbols-not-imported
```
The error observed was:
``` 
rust/build/x86_64-unknown-linux-gnu/stage1/bin/rustc: error while loading shared libraries: librustc_driver-8fd070a3dbd486b2.so: cannot open shared object file: No such file or directory
```

Providing the LD_LIBRARY_PATH for compiling and running the tests fixes the issue. Similar fix from the past - https://github.com/rust-lang/rust/pull/113889 for a similar error.

@jethrogb @raoulstrackx

### CODE REVIEW

File: src/tools/compiletest/src/runtest.rs [hunk 1]
```diff
@@ -3784,6 +3784,7 @@ impl<'test> TestCx<'test> {
             .env("TMPDIR", &tmpdir)
             .env("LD_LIB_PATH_ENVVAR", dylib_env_var())
             .env("HOST_RPATH_DIR", cwd.join(&self.config.compile_lib_path))
+            .env("LD_LIBRARY_PATH", cwd.join(&self.config.compile_lib_path))
```
Comment (jieyouxu) [inline src/tools/compiletest/src/runtest.rs:3787]: Hmm, I wonder if this is the right place to solve it. The original run-make specifies `LD_LIBRARY_PATH` via `LD_LIB_PATH_ENVVAR` (when that is the correct `LD_LIB_PATH_ENVVAR` for the target):

https://github.com/rust-lang/rust/blob/7f3a62b17b3d72746c8c839baca851bac7544180/src/tools/compiletest/src/runtest.rs#L3582

In `tools.mk`, then `LD_LIBRARY_PATH` (which is the suitable `LD_LIB_PATH_ENVVAR`) is set to the correct paths

https://github.com/rust-lang/rust/blob/536606bc5df7e6045c6dbedc2f852ba3d38581f1/tests/run-make/tools.mk#L3-L6

My guess is that

https://github.com/rust-lang/rust/blob/536606bc5df7e6045c6dbedc2f852ba3d38581f1/src/tools/run-make-support/src/lib.rs#L142-L154

is broken, but this should be the correct place to fix it. In particular, I think you'd just need to preserve the semantics of

```make
 HOST_RPATH_ENV = \ 
     $(LD_LIB_PATH_ENVVAR)="$(TMPDIR):$(HOST_RPATH_DIR):$($(LD_LIB_PATH_ENVVAR))" 
 TARGET_RPATH_ENV = \ 
     $(LD_LIB_PATH_ENVVAR)="$(TMPDIR):$(TARGET_RPATH_DIR):$($(LD_LIB_PATH_ENVVAR))"
```

Comment (jieyouxu) [reply src/tools/compiletest/src/runtest.rs:3787]: Ah, I think `"TARGET_RPATH_ENV"` should instead be `TARGET_RPATH_DIR`? Can you check if that fixes the issue?

Comment (aditijannu) [reply src/tools/compiletest/src/runtest.rs:3787]: @jieyouxu Updating TARGET_RPATH_ENV with TARGET_RAPTH_DIR doesn't work on sgx.

Comment (jieyouxu) [reply src/tools/compiletest/src/runtest.rs:3787]: Hmm nevermind then. I see the original run-make test also had this issue...

Comment (jieyouxu) [reply src/tools/compiletest/src/runtest.rs:3787]: Actually, could you try changing

```rs
- for p in env::split_paths(&env::var("TARGET_RPATH_ENV").unwrap()) { 
-     paths.push(p.to_path_buf()); 
- }
+ for p in env::split_paths(&env::var("HOST_RPATH_DIR").unwrap()) { 
+     paths.push(p.to_path_buf()); 
+ }
+ for p in env::split_paths(&env::var("TARGET_RPATH_DIR").unwrap()) { 
+     paths.push(p.to_path_buf()); 
+ }
```

Judging from https://github.com/rust-lang/rust/pull/113889, `HOST_RPATH_DIR` should have the paths you need (I hope)

Comment (Mark-Simulacrum) [reply src/tools/compiletest/src/runtest.rs:3787]: Fixing this with something like this makes sense to me.

**Comment by aditijannu** [comment]
Fixed in https://github.com/rust-lang/rust/pull/123763

**Pull request closed without merging**